### PR TITLE
Added option to add rotated Box obstacle to TileCache.

### PR DESCRIPTION
**Pull request by aymarfisherman**
Added option to add rotated Box obstacle to TileCache (without messing with the existing box).

Very useful when you have a obstacle with a big difference in x to z ratio and that could be rotated (not aligned), like a wall.

### CODE REVIEW

File: DetourTileCache/Include/DetourTileCache.h [hunk 1]
```diff
@@ -38,6 +38,7 @@ enum ObstacleState
 enum ObstacleType
 {
 	DT_OBSTACLE_CYLINDER,
+	DT_OBSTACLE_AABB,
```
Comment (jakobbotsch) [inline DetourTileCache/Include/DetourTileCache.h]: Please rename this back and use a new name for your addition.
For example: `DT_OBSTACLE_ORIENTED_BOX`.
To make it more clear you can add some comments that `DT_OBSTACLE_BOX` is an AABB and `DT_OBSTACLE_ORIENTED_BOX` is an OBB.

File: DetourTileCache/Include/DetourTileCache.h [hunk 2]
```diff
@@ -48,18 +49,26 @@ struct dtObstacleCylinder
 	float height;
 };
 
-struct dtObstacleBox
+struct dtObstacleAabb
```
Comment (jakobbotsch) [inline DetourTileCache/Include/DetourTileCache.h]: Undo this renaming please.

File: DetourTileCache/Include/DetourTileCache.h [hunk 3]
```diff
@@ -48,18 +49,26 @@ struct dtObstacleCylinder
...
 	float bmin[ 3 ];
 	float bmax[ 3 ];
 };
 
+struct dtObstacleBox
+{
+	float center[ 3 ];
+	float halfextents[ 3 ];
+	float rotaux[ 2 ]; //{ cos(0.5f*angle)*sin(-0.5f*angle); cos(0.5f*angle)*cos(0.5f*angle) - 0.5 }
+};
```
Comment (jakobbotsch) [inline DetourTileCache/Include/DetourTileCache.h]: Please call these `halfExtents` and `rotAux`.

File: DetourTileCache/Include/DetourTileCache.h [hunk 4]
```diff
@@ -48,18 +49,26 @@ struct dtObstacleCylinder
...
+
 static const int DT_MAX_TOUCHED_TILES = 8;
 struct dtTileCacheObstacle
 {
 	union
 	{
 		dtObstacleCylinder cylinder;
+		dtObstacleAabb aabb;
```
Comment (jakobbotsch) [inline DetourTileCache/Include/DetourTileCache.h]: Rename this back and call the new one something like `orientedBox` or `obox`.

**Comment by aymarfisherman** [comment]
I'm glad I could contribute.
I did what you requested.
Let me know if you spot anything else I could do to improve this PR.

File: DetourTileCache/Include/DetourTileCache.h [hunk 5]
```diff
@@ -130,8 +139,14 @@ class dtTileCache
...
 	dtStatus removeTile(dtCompressedTileRef ref, unsigned char** data, int* dataSize);
 	
+	// Cylinder obstacle.
 	dtStatus addObstacle(const float* pos, const float radius, const float height, dtObstacleRef* result);
+
+	// Aabb obstacle.
 	dtStatus addBoxObstacle(const float* bmin, const float* bmax, dtObstacleRef* result);
+
+	// Obb obstacle: can be rotated in Y.
+	dtStatus addBoxObstacle(const float* bmin, const float* bmax, const float yRadians, dtObstacleRef* result);
```
Comment (jakobbotsch) [inline DetourTileCache/Include/DetourTileCache.h]: `bmin`/`bmax` here should be `center` and `halfExtents`. Though can we rename `halfExtents` to just `extents`? This is the convention used in other places (eg `findNearestPoly`). To make it more clear, here is an example: assuming the rotation is 0, passing center = `0, 0, 0` and `extents = 1, 1, 1` should mark the box from `-1, -1, -1` to `1, 1, 1`.

File: DetourTileCache/Include/DetourTileCacheBuilder.h [hunk 6]
```diff
@@ -130,6 +130,9 @@ dtStatus dtMarkCylinderArea(dtTileCacheLayer& layer, const float* orig, const fl
 dtStatus dtMarkBoxArea(dtTileCacheLayer& layer, const float* orig, const float cs, const float ch,
 					   const float* bmin, const float* bmax, const unsigned char areaId);
 
+dtStatus dtMarkBoxArea(dtTileCacheLayer& layer, const float* orig, const float cs, const float ch,
+					   const float* center, const float* halfExtents, const float* rotAux, const unsigned char areaId);
```
Comment (jakobbotsch) [inline DetourTileCache/Include/DetourTileCacheBuilder.h]: `extents` please, as used in other places in Detour.

File: DetourTileCache/Include/DetourTileCache.h [hunk 7]
```diff
@@ -54,13 +55,21 @@ struct dtObstacleBox
 	float bmax[ 3 ];
 };
 
+struct dtObstacleOrientedBox
+{
+	float center[ 3 ];
+	float halfExtents[ 3 ];
```
Comment (jakobbotsch) [inline DetourTileCache/Include/DetourTileCache.h]: Just `extents` please (see my comment below).

File: DetourTileCache/Source/DetourTileCache.cpp [hunk 8]
```diff
@@ -420,6 +420,45 @@ dtStatus dtTileCache::addBoxObstacle(const float* bmin, const float* bmax, dtObs
 	return DT_SUCCESS;
 }
 
+dtStatus dtTileCache::addBoxObstacle(const float* center, const float* halfExtents, const float yRadians, dtObstacleRef* result)
```
Comment (jakobbotsch) [inline DetourTileCache/Source/DetourTileCache.cpp]: `extents`

File: DetourTileCache/Source/DetourTileCacheBuilder.cpp [hunk 9]
```diff
@@ -2041,6 +2041,60 @@ dtStatus dtMarkBoxArea(dtTileCacheLayer& layer, const float* orig, const float c
 	return DT_SUCCESS;
 }
 
+dtStatus dtMarkBoxArea(dtTileCacheLayer& layer, const float* orig, const float cs, const float ch,
+					   const float* center, const float* halfExtents, const float* rotAux, const unsigned char areaId)
```
Comment (jakobbotsch) [inline DetourTileCache/Source/DetourTileCacheBuilder.cpp]: `extents`

**Comment by aymarfisherman** [comment]
I changed it to 'extents' and fixed the argument naming mismatch between .h and .cpp.

Though I'd like to leave a suggestion: 

Please, consider renaming all 'extents' to 'halfExtents', since 'extents' is misleading and ambiguous, especially when 'halfExtents' is  widely used in the industry. It's expected that people will pass 2x'halfExtents' if you ask for 'extents'.

It happened to me and maybe to other people: I thought 'extents' were the actual the full 'extents' of the box and I was wrongly using the library for years, passing 2.0*halfExtents everywhere it asked for 'extents'. I found out by accident (when implementing this PR) that the 'extents' variable actually meant 'halfExtents'.

If not for those reasons, you should still consider this change for shear disambiguation.

**Comment by jakobbotsch** [comment]
Thanks! I am not opposed to renaming `extents` to `halfExtents` everywhere in the library. Would you mind submitting a PR for that?

**Pull request merged**